### PR TITLE
Add HTML tracing capability

### PR DESCRIPTION
**Pull request by sauyon**

### CODE REVIEW

**Comment by sauyon** [comment]
I'm rather confused as to why the tests are failing; they seem to work fine on my local machine.

**Comment by sauyon** [comment]
Derp, nevermind: #474

**Comment by smowton** [comment]
Weird base branch on this -- retarget to main?

**Comment by smowton** [comment]
In fact, turned out I had the power to do that :)

File: codeql-tools/pre-finalize.cmd [hunk 1]
```diff
@@ -0,0 +1,19 @@
+@echo off
+SETLOCAL EnableDelayedExpansion
+
+rem Some legacy environment variables for the autobuilder.
+set LGTM_SRC=%CD%
```
Comment (smowton) [inline codeql-tools/pre-finalize.cmd:5]: Should this also be set in the `.sh` script?

File: extractor/dbscheme/dbscheme.go [hunk 2]
```diff
@@ -362,6 +362,15 @@ func NewUnionType(name string, parents ...*UnionType) *UnionType {
 	return tp
 }
 
+// AddChild adds the type with given `name` to the union type.
+// This is useful if a type defined in a snipped should be a child of a type defined in Go.
```
Comment (smowton) [inline extractor/dbscheme/dbscheme.go:366]: in a snipped?

File: extractor/dbscheme/tables.go [hunk 3]
```diff
@@ -44,12 +44,90 @@ snapshotDate(unique date snapshotDate : date ref);
 sourceLocationPrefix(varchar(900) prefix : string ref);
 `)
 
+var xmlSnippet = AddDefaultSnippet(`
+/*
+ * XML Files
```
Comment (smowton) [inline extractor/dbscheme/tables.go:49]: Perhaps note in the comment that this is identical to xyz codeql schema

File: ql/src/semmle/go/HTML.qll [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+   * An HTML element.
+   *
+   * Example:
+   *
+   * ```
+   * <a href="semmle.com">Semmle</a>
+   * ```
+   */
+  class Element extends Locatable, @xmlelement {
+    Element() { exists(HtmlFile f | xmlElements(this, _, _, _, f)) }
```
Comment (smowton) [inline ql/src/semmle/go/HTML.qll:23]:
```suggestion
    Element() { xmlElements(this, _, _, _, any(HtmlFile f)) }
```

File: ql/src/semmle/go/HTML.qll [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+    override Location getLocation() { xmllocations(this, result) }
+
+    /**
+     * Gets the name of this HTML element.
+     *
+     * For example, the name of `<br>` is `br`.
```
Comment (smowton) [inline ql/src/semmle/go/HTML.qll:30]:
```suggestion
     * For example, the name of `<br/>` is `br`.
```

File: ql/src/semmle/go/HTML.qll [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+    /**
+     * Gets an attribute of this element that has the given name.
+     */
+    Attribute getAttributeByName(string name) {
+      result = getAnAttribute() and
+      result.getName() = name
+    }
+
+    TextNode getTextNode() { result.getParent() = this }
```
Comment (smowton) [inline ql/src/semmle/go/HTML.qll:82]: Doc comment

File: ql/src/semmle/go/frameworks/GoproxyGoproxy.qll [hunk 7]
```diff
@@ -0,0 +1,6 @@
+/** Provides models of commonly used functions and types in the [github.com/goproxy/goproxy](https://pkg.go.dev/github.com/goproxy/goproxy) package. */
```
Comment (smowton) [inline ql/src/semmle/go/frameworks/GoproxyGoproxy.qll:1]: Accidentally added to this PR?

File: ql/test/extractor-tests/html/test.html [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<h1>{{.PageTitle}}</h1>
```
Comment (smowton) [inline ql/test/extractor-tests/html/test.html:1]: Maybe add a comment indicating this is an example of Go's HTML templating, and intentionally a fragment

**Comment by smowton** [comment]
Perhaps this would be a good place to put an evaluation ticklist, since I'm finding it hard to keep a handle on what's done and what's outstanding on that front

File: ql/src/semmle/go/HTML.qll (hunk 5 above)
Comment (sauyon) [reply ql/src/semmle/go/HTML.qll:30]: I believe `<br>` is still technically correct in HTML, unless the XHTML purists have won.

**Comment by sauyon** [comment]
> Weird base branch on this -- retarget to main?

I tend to use base branches when I have PRs that depend on each other for ease of review.

File: codeql-tools/pre-finalize.cmd (hunk 1 above)
Comment (sauyon) [reply codeql-tools/pre-finalize.cmd:5]: I think it's probably not necessary and I overzealously copied the batch code without thinking about it.

**Comment by sauyon** [comment]
Merged as part of #473

**Pull request closed without merging**